Remote Data Converter & PayloadCodec (#1330)
temporalio/sdk-java: temporal-remote-data-encoder/README.md

# Temporal Remote Data Encoder module

## What problems does Remote Data Encoder (RDE) solves?

> - Allows to reuse complicated encryption logic written once between different languages
> - Allows tctl to encode payloads for tctl workflow start and Temporal WebUI to decode encrypted payloads
> - Allows creation of a service that has an access to encryption keys and performing the encryption/decryption instead of allowing developer workstations to access the keys directly.

Reviewer: ```suggestion
- Allows reuse of complicated encryption logic written once between different languages
- Allows `tctl` to encode payloads for `tctl workflow start` and Temporal WebUI to decode encrypted payloads
- Allows creation of a service that has an access to encryption keys for performing the encryption/decryption instead of a developer workstation or service accessing the keys directly.
```

## How does RDE work?

### RDE Server

- It's an Http Server responding to POST requests with 'application/json' content type and sending back responses with 'application/json' content type

> - It listens on some base url which can be postfixed with /encode or /decode for encoding and decoding respectively
> - Expect and emit `io.temporal.api.common.v1.Payloads` serialized to json format using [Proto3 Json Mapping](https://developers.google.com/protocol-buffers/docs/proto3#json) and sent in Http Request / Response body

Reviewer: ```suggestion
- It listens on some base url which is postfixed with /encode or /decode for encoding and decoding respectively
- Expects and emits `io.temporal.api.common.v1.Payloads` serialized to json format using [Proto3 Json Mapping](https://developers.google.com/protocol-buffers/docs/proto3#json) and sent in Http Request / Response body
```

- `io.temporal.rde.servlet.RDEServlet4` - Servlet compatible with 3.0 and 4.0 standard Java Servlet Specification that can be deployed to any application or servlet container supporting Servlet Specification of version 3.0 or 4.0

### RDE Codec

Can be used as one of the codecs in `io.temporal.common.converter.CodecDataConverter` configured on `WorkflowClientOptions#dataConverter` to use Remote Data Encoder Server for payloads encoding.
Please keep in mind that this is an optional component. If you use RDE Server to decode data for tctl or WebUI purposes, you don't have to use it for encoding/decoding of payloads on Workflow Client or Workers side. They may perform the encoding locally.

This module provides a reference implementation of RDE codec that should be used by users as a base for their own RDE codecs: `io.temporal.payload.codec.AbstractRemoteDataEncoderCodec` by implementing a POST method using the HTTP client of their choice.
This module may supply some standard implementations for popular HTTP Clients, including `io.temporal.payload.codec.OkHttpRemoteDataEncoderCodec` for [OkHttpClient](https://square.github.io/okhttp/).

## Security notes

> Existence of RDE in a system creates an additional party that may have an access to the encryption keys and exposing APIs to encode and decode any payloads. Users should evaluate and mitigate potential security risks of this solution.

Reviewer: This is true sometimes, but not always. You could actually use it to _reduce_ the number of places keys exist by routing all encoding and decoding through it, but that's just a different tradeoff.

Author: Added clarifications. It reduces number of parties having access to keys, but introduces a completely new thing to protect access to (other that keys).
